Issue #213: checkstyle-tester: support regression on new checks/tokens

MEZk opened the pull request:
issue #213 

1) Added ability to generate Checkstyle reports for patch branch only with diff.groovy
2) Extended README.md
3) diff.groovy does not have ```-i``` command line option (only launch groovy has), so it was removed from Travis CI yml.

Further refactoring and performance optimization of diff.groovy will be performed in the scope of https://github.com/checkstyle/contribution/issues/182.

diff.groovy should be renamed to reports.groovy or something similar in separate issue.

@rnveach
Maybe we need to have the manual for launch.groovy in a separate file. Do contributors need to know about it from the main REAMDE.md of checkstyle-tester?

rnveach commented:
> diff.groovy should be renamed to reports.groovy 

Any reason why? I think it makes more sense as it does more than create the reports, it generates the differences.

> Do contributors need to know about it from the main REAMDE.md of checkstyle-tester?

Would it make more sense for each utility to have it's own readme file?
If not, I am fine with it staying in the same file as we can link directly to diff.groovy.

rnveach commented on checkstyle-tester/diff.groovy:
```diff
@@ -161,6 +135,20 @@ def hasUnstagedChanges(gitRepo) {
 
+def runPlainCheckstyleReportsGeneration(cfg) {
+    generateCheckstyleReport(cfg.localGitRepo, cfg.patchBranch, cfg.patchCheckstyleCfg, cfg.listOfProjects, cfg.reportsDir)
+    generateSummaryIndexHtml(cfg.reportsDir)
```
@MEZk In the end, I only want to look at diff reports and not the summary report. :)
I think our report is alot easier to view.

If we made `patch-diff-report-tool` support no base config or base reports, we wouldn't need this extra parsing or different code for the summary index. Then all reports would look the same and it would be easier to make more additions to the summary, if we decide to.

Does this sound like a better idea?
@romani What do you think?

MEZk replied:
@rnveach 
> In the end, I only want to look at diff reports and not the summary report. :)
I think our report is alot easier to view.

If I generate reports for more than one project, I think that having summary page makes it easier to review the reports which are uploaded to github.io.

> If we made patch-diff-report-tool support no base config or base reports, we wouldn't need this extra parsing or different code for the summary index. Then all reports would look the same and it would be easier to make more additions to the summary, if we decide to.

Agree.

rnveach replied:
>  I think that having summary page 

I don't know how I keep mixing things up. 
When I said summary report, I meant the normal `launch.groovy` report. I'm sorry again.

http://rveach.no-ip.org/checkstyle/regression/reports/20/checkstyle/index.html I meant this report.

rnveach replied:
PR for patch-diff-report-util started at https://github.com/checkstyle/contribution/pull/216

MEZk commented:
@rnveach 

> Would it make more sense for each utility to have it's own readme file?
If not, I am fine with it staying in the same file as we can link directly to diff.groovy.

I guess, yes. Contributor who wants to generate diff reports should not know about launch.groovy. However, those who want to fix the issues of launch.groovy and checkstyle-tester project itself, should. That is why I think it is better to have separate readme files.

romani commented:
if we so care about new contributors and we do not want to support two separate scripts

we need to rename `diff.groovy` to `cs-tester.groovy` and make separate modes for diff and single launch.

rnveach commented:
@MEZk PR for `patch-diff-report` was merged. We can continue this PR based on new comments romani made in issue.
I'm not sure which issue this will be filed under yet.

MEZk commented:
@rnveach 
Updated.

@romani 
The build failed due to Codenarc's violaions. I added 2 def tokens as I still thing that we should use the Groovy's keyword. Please, merge https://github.com/checkstyle/contribution/pull/220 to let me restart the build and get the actual Codenarc's violations.

rnveach commented on checkstyle-tester/README.md:
```diff
@@ -1,138 +1,59 @@
-**baseBranch** (b) - name of the base branch in local Checkstyle repository (optional, default is master branch);
+**baseBranch** (b) - name of the base branch in local Checkstyle repository (optional, if asent, then the tool will use only patchBranch in case patchOnly is set, otherwise baseBrach will be set to 'master');
+
+**patchOnly** (po) - whether to generate report only for the patch branch. If the option is set, then baseBranch, baseConfig and config should not be specified as the tool will finish the execution with the error. You must specify patchBranch and patchConfig if you enable this option. (optional, not specified by default)
```
romani wants the option to be named `mode` with the values `diff` (default) or `single`.
https://github.com/checkstyle/contribution/issues/213#issuecomment-297531342

I think we should explicitly tell users why they would use this option.
This is for any option that can't work on master/base branch, which are new properties, new tokens, or new modules.

MEZk replied:
Done

rnveach commented on checkstyle-tester/README.md:
```diff
@@ -152,7 +73,7 @@ When the script finishes its work the following directory structure will be crea
 */reports/diff* - directory with diff reports;
 
-*reports/baseBranch* - directory with Checkstyle reports which are generated with Checkstyle base version (based on specified base branch);
+*reports/baseBranch* - directory with Checkstyle reports which are generated with Checkstyle base version (based on specified base branch. If patchOnly is enabled then the diractory will not be created.);
```
`diractory` -> `directory`

MEZk replied:
Done

rnveach commented on checkstyle-tester/diff.groovy:
```diff
@@ -103,19 +109,20 @@ def areValidCliOptions(cliOptions) {
             || config && baseConfig
             || config && patchConfig && baseConfig) {
-        err.println "Error: you should specify either \'config\' or \'baseConfig\' and \'patchConfig\'!"
+        err.println "Error: you should specify either \'config\' or \'baseConfig\' and \'patchConfig\', or \'patchConfig\' only!"
```
I think there should be a comma before first 'or' too to more clearly make the separation from the 3 items.

MEZk replied:
Done.

rnveach commented on checkstyle-tester/diff.groovy:
```diff
@@ -103,19 +109,20 @@ def areValidCliOptions(cliOptions) {
+    if (baseConfig && !patchConfig
+            || config && patchConfig
             || config && baseConfig
             || config && patchConfig && baseConfig) {
```
unrelated to changes but how can this or statement be hit while the others aren't?
Whenever this line is `true` previous 2 lines are `true` as well.

MEZk replied:
Removed unnecessary condition.

rnveach commented on checkstyle-tester/diff.groovy:
```diff
@@ -66,11 +70,12 @@ def getCliOptions(args) {
         b(longOpt: 'baseBranch', args: 1, required: false, argName: 'branch_name', 'Base branch name. Default is master (optional, default is master)')
         p(longOpt: 'patchBranch', args: 1, required: true, argName: 'branch_name', 'Name of the patch branch in local git repository (required)')
-        bc(longOpt: 'baseConfig', args: 1, required: false, argName: 'path', 'Path to the base checkstyle config file (required if patchConfig is specified)')
+        bc(longOpt: 'baseConfig', args: 1, required: false, argName: 'path', 'Path to the base checkstyle config file (optional, optional, if asent then the tool will use only patchBranch in case patchOnly is set, otherwise baseBrach will be set to \'master\')')
```
`asent` -> `absent`
`baseBrach` -> `baseBranch`

MEZk replied:
Fixed

rnveach commented on checkstyle-tester/README.md:
```diff
@@ -1,138 +1,61 @@
-**baseBranch** (b) - name of the base branch in local Checkstyle repository (optional, default is master branch);
+**baseBranch** (b) - name of the base branch in local Checkstyle repository (optional, if asent, then the tool will use only patchBranch in case the tool mode is 'single', otherwise baseBrach will be set to 'master');
+
+**mode** (m) - The mode of the tool: 'single' or 'diff'. Set this option to 'single' if your patch branch contains changes for any option that can't work on master/base branch. For example, for new properties, new tokens, or new modules.
```
I don't see a mention if this is required or optional.
We should mention that `diff` is the default?

After the sentence "For example, ...", please add a sentence that says something like "For all other changes, 'diff' mode should be the preferred mode used." so it is clear.

MEZk replied:
> I don't see a mention if this is required or optional.
We should mention that diff is the default?

It was mentioned at the end of the property description. I moved it at the start of the option description.

> After the sentence "For example, ...", please add a sentence that says something like "For all other changes, 'diff' mode should be the preferred mode used." so it is clear.

Done.

rnveach commented:
@MEZk #220 was merged. Please rebase and make CI green.
@romani Please do final review.

MEZk commented:
@rnveach 
Rebased.

@romani 
I still think that we should use the ```def``` keyword.

romani commented:
> I still think that we should use the def keyword.

lets vote. I presume there 3 of us this project.
@rnveach , please share your position

My position is to do not use "def".

MEZk commented:
@romani 
In your previous comment you suggest rewriting the code in the way it will become more similar to Java. This will mean that we should also prohibit the usage of Groovy's closures and other language structures and make Groovy code exactly the same as Java code.

rnveach commented:
@romani I am in favor of the `def` keyword to make defining variables stand out, like defining the type in Java.

romani commented:
ok,  2 against 1 , @MEZk , please disable a rule with wide explanation.

MEZk commented:
@romani @rnveach 
According to [Groovy Official Style guide](http://groovy-lang.org/style-guide.html):
> I’ll finish on some words on when and how to use optional typing. Groovy lets you decide whether you use explicit strong typing, or when you use def. I’ve got a rather simple rule of thumb: whenever the code you’re writing is going to be used by others as a public API, you should always favor the use of strong typing, it helps making the contract stronger, avoids possible passed arguments type mistakes, gives better documentation, and also helps the IDE with code completion. Whenever the code is for your use only, like private methods, or when the IDE can easily infer the type, then you’re more free to decide when to type or not.

As we design the tool which will be used and supported by other contributors, maybe it is even better to use strong typing. What do you think?

romani commented:
my position is not use "def". But I am fine with "def" if most of maintainers want it.

MEZk commented:
@romani 
Ok, lets leave def for now. All in all, style guide is a personal opinion of the man who wrote it, we can take into account his opinion, but still write code as we believe will be better.

romani commented:
CI has to be Green.

MEZk commented:
@romani 
The rule 'ExplicitCallToEqualsMethod' should be disabled as well, as it makes code cryptic. In Groovy ```==``` is an overloaded operator which invokes ```equasl```, but it is weird for Java contributors. I disable the rule.

romani commented:
Ok to disable this rule.
http://codenarc.sourceforge.net/codenarc-rules-groovyism.html

romani commented:
@MEZk, what stop you from running codenarc on local ? To send update that will pass finally.

>"The line exceeds 120"

Please fix this in this PR, as you will change code again. We have 15 violations like this.

MEZk commented:
@romani 
CIs are fixed.
